rfcs/20230118-f16-accumulation/README.md

```diff
@@ -0,0 +1,125 @@
+implementations based on propagation_kind. We could rely on this
+mecanism to guard dispatching of implementations using f16
+accumulation with the two following caveats:
+- this would increase validation costs as we would need to validate both builds
+- it adds development burden. Currently we don't know any information
+  on the accumulation datatype for each implementations. Properly
+  guarding implementation dispatch would either need to be scattered
+  in implementation using f16 accumulation (which is brittle), or
+  would require to add internal APIs to query implementations for their
+  accumulation datatype.
+
+## A new `accumulation_datatype` attribute (recommended)
+
+The last option is to rely on a new primitive attribute to get a hint
+on minimal accuracy required in accumulators. There are multiple
+options for what values it should accept.
+
+For all those options, note that implementations would still be able
+to use a different datatype internally if the result is unchanged and
+it is faster (e.g. using `f32` intermediate accumulators when
+accumulation datatype is `f16`, or in some `s8/u8` computations when
+it yields the same result).
+
+### Use a datatype (recommended).
+
+In those cases, the user will have to explicitely pass `f32`, `f16`,
+`s32`, ...  When the user does not set the attribute, we will use a
+default value of `undef`, which would allow all implementation to be
+dispatched, including those with `f16` accumulation.
+
+Pros:
+- no new datatype introduced
+- the user explicitly passes the datatype
+
+Cons:
+- not fully clear from API that it is a hint
+
+### Use a new accumulation mode
+
+Here, we would define a new datatype
+
+```c++
+enum class accumulation_mode {
+    /// Default behavior, uses accumulators with 32-bit precision 
+    accurate = dnnl_accumulation_mode_accurate,
+    /// Might use lower accuracy accumulators that were tested to work for most models
+    fast = dnnl_accumulation_mode_fast,
+};
+```
+
+The pros and cons are the exact opposite of using datatypes.
+
+Pros:
+- Clear from API that it is a hint
+Cons:
+- the user does not know exactly which datatype is used for accumulation
+- it introduces another datatype.
+
+### A note on the default behavior.
+
+Note that for both options, we can use the `ONEDNN_ENABLE_WORKLOAD` to
+set change the default behavior.  For the datatype flavor, `undef`
+would enable `f16` accumulation dispatch for
+`ONEDNN_ENABLE_WORKLOAD=INFERENCE` but not for
+`ONEDNN_ENABLE_WORKLOAD=TRAINING`.  For the mode flavor, default would
+be `fast` for `ONEDNN_ENABLE_WORKLOAD=INFERENCE`, and accurate for
+``ONEDNN_ENABLE_WORKLOAD=TRAINING`.
+
+This would allow to run full validation on a single build, and let the
+current openVINO/ONNXruntime integrations unchanged, as they use
+`ONEDNN_ENABLE_WORKLOAD=INFERENCE`.  Though it will require extra
+validation still (build + interface test for default value).
+
+If we elect to go with a fixed default behavior, openVINO and
```

Comment:
I might be wrong, but I think I've heard that TF always set prop as training, so this optimization based on prop might night work for TF.

Reply:
Yes this is correct, most framework actually always use training propagation kinds, as they cannot assume inference unless weights are constant/frozen (which is often not the case in practice).

Though the primary users of fp16 accumulation are deployment toolkits (openVINO/ONNXruntime), and for those, they can and do use inference propagation kind when appropriate.